x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java

```diff
@@ -269,17 +274,74 @@ public boolean equals(Object obj) {
             TopMetric other = (TopMetric) obj;
             return sortFormat.equals(other.sortFormat)
                     && sortValue.equals(other.sortValue)
-                    && Arrays.equals(metricValues, other.metricValues);
+                    && metricValues.equals(other.metricValues);
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(sortFormat, sortValue, Arrays.hashCode(metricValues));
+            return Objects.hash(sortFormat, sortValue, metricValues);
         }
 
         @Override
         public String toString() {
-            return "TopMetric[" + sortFormat + "," + sortValue + "," + Arrays.toString(metricValues) + "]"; 
+            return "TopMetric[" + sortFormat + "," + sortValue + "," + metricValues + "]"; 
+        }
+    }
+
+    static class MetricValue implements Writeable, ToXContent {
+        private final DocValueFormat format; 
+        private final SortValue value;
+
+        MetricValue(DocValueFormat format, SortValue value) {
+            this.format = format;
+            this.value = value;
+        }
+
+        DocValueFormat getFormat() {
+            return format;
+        }
+
+        SortValue getValue() {
+            return value;
+        }
+
+        MetricValue(StreamInput in) throws IOException {
+            format = in.readNamedWriteable(DocValueFormat.class);
+            value = in.readNamedWriteable(SortValue.class);
+        }
+
+        Number numberValue() {
+            return value.numberValue();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeNamedWriteable(format);
+            out.writeNamedWriteable(value);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            return value.toXContent(builder, format);
+        }
+
+        @Override
+        public String toString() {
+            return format + "," + value;
```

Comment:
```suggestion
            return "MetricValue(" + format + "," + value + ")";
```
Pet peeve of mine, unless we have a good reason not to, we should include the class name in the `toString`.  I've gotten burned by this a few times trying to write exception messages.